test/integration/scheduler_perf/util.go

```diff
@@ -199,10 +199,6 @@ func collectHistogram(metric string, labels map[string]string) *DataItem {
 	q99 := hist.Quantile(0.95)
 	avg := hist.Average()
 
-	// clear the metrics so that next test always starts with empty prometheus
-	// metrics (since the metrics are shared among all tests run inside the same binary)
-	hist.Clear()
```

Comment:
shouldn't we still call Reset here?

Reply:
Not really, Reset() or Clear() doesn't behave as we expected.

Reply:
How can we prevent metrics from leaking into the next test then?

Reply:
We now have the new logic in test/integration/scheduler_perf/scheduler_perf_test.go:

```go
if !concreteOp.SkipWaitToCompletion {
    // SkipWaitToCompletion=false indicates this step have waited for the Pods to be scheduled.
    // So we reset the metrics in global registry; otherwise metrics gathered in this step/test
    // will be rolled over to next step/test.
    legacyregistry.Reset()
}
```

Reply:
but that's conditional.... better safe to always remove

Reply:
> better safe to always remove

An unconditional reset doesn't make a lot of sense, as during the asynchronous (SkipWaitToCompletion=true) createPods operation, we cannot really get rid of their metrics. This is also mentioned in https://github.com/kubernetes/kubernetes/issues/99218#issuecomment-782297367. So I choose to only reset when it's a synchronous operation.

Reply:
As @alculquicondor mentioned, let's assume this configuration (`skipWaitToCompletion` is `false` in all cases):
```
- name: Unschedulable
  workloadTemplate:
  - opcode: createNodes
    countParam: $initNodes
  - opcode: createPods
    countParam: $measurePods
    podTemplatePath: config/pod-default.yaml
    collectMetrics: true
  workloads:
  - name: 500Nodes/1000Pods
    params:
      initNodes: 500
      measurePods: 1000
  - name: 5000Nodes/5000Pods
    params:
      initNodes: 5000
      measurePods: 5000
```
Once the second workload (`5000Nodes/5000Pods`) gets ran, metrics collected from the previous workload will get counted as well. I second what @alculquicondor says. `Reset()` needs to be ran at every end of `for opIndex, op := range tc.WorkloadTemplate` loop at least.

Reply:
> as during the asynchronous (SkipWaitToCompletion=true) createPods operation, we cannot really get rid of their metrics

I see the point. In which case we need to wait until all pods get scheduled before the next workload gets processed.

Reply:
I agree, we should reset at the end of every workload but that is implicitly guaranteed because we don't have any workloadTemplate that uses SkipWaitToCompletion as the last step (which doesn't really make sense). perhaps we should add guards to make SkipWaitToCompletion in the last step isn't possible.

Reply:
@adtac is right, however, we probably don't want to ban the pattern using SkipWaitToCompletion at the last step, so adding a double-check at the end of each workload's running is fine with me. Updated the code.

> I second what @alculquicondor says. `Reset()` needs to be ran at every end of `for opIndex, op := range tc.WorkloadTemplate` loop at least.

Technically correct, but to save cost, we only need to run Reset() when SkipWaitToCompletion is false.

Reply:
Current logic for resetting lgtm